### ISSUE TITLE
"Loss: nan" in training

### ISSUE THREAD

**Issue by 0x7o** [ai-forever/mgpt#1]
When trying to teach mGPT through a [notebook](https://colab.research.google.com/drive/1qkDhzEab2MXvohOuQYgKixHHimlh1Oh2?usp=sharing), after a while loss equals "nan".
![image](https://user-images.githubusercontent.com/96177948/164456572-9dbc3062-268a-4fa3-8327-deb90993c522.png)
In this case, the generation is disrupted:
![image](https://user-images.githubusercontent.com/96177948/164456179-c6ba0812-496e-4ae1-94bb-c8bdbecf7fa1.png)
File size for learning - 45 mb
Learning Rate - 1e-5
GPU - Nvidia Tesla P100

**Comment by ollmer** [comment]
Hi! Unfortunately, the amount of GPU memory of colab instances is very limited, so to be able to fit tuning example in it we convert model to an fp16 there and use SGD optimizer that not requires additional memory for parameters. But this tuning mode is definitely far from stable and can lead to a NaN loss due to limitations of fp16 numeric range. We're planning to make an updated version of colab tuning with more advanced techniques to reduce memory consumption. This could possibly allows us to use mixed-precision training with this memory constrains.
For now you can try to tune model [using notebook from this repo](https://github.com/ai-forever/mgpt/blob/main/notebooks/mgpt_tuning_example.ipynb) on your own GPU, it uses float32 training and AdamW optimizer but requires 32Gb of GPU RAM.

**Comment by ollmer** [comment]
Recent experiment shows that model with gradient checkpointing and SGD optim fits into 16Gb of GPU RAM.
Try this model initialization:
`model = GPT2LMHeadModel.from_pretrained("sberbank-ai/mGPT", gradient_checkpointing=True, use_cache=False)`, but remove further conversion `model = model.half()`
For me it consumes 15774Mb of RAM with batch size 1, seq length 512.

**Comment by 0x7o** [comment]
Thank you! It really worked.
![image](https://user-images.githubusercontent.com/96177948/164682645-85150576-d6fb-487c-8f3e-651c916e0e59.png)
